### ISSUE TITLE
Backend data structuring

### ISSUE THREAD

**Issue by Gryphire** [Gryphire/wisperbot#16]
For the research on Wisperbot, we'll need to make sure that backend data is collected. In principle, we would need to collect _all of the behaviours_ that take place within a chat with Wisperbot (i.e., logging users' behaviours as well as Wisperbot's behaviours). Voice messages and their transcripts are already being saved (not fed back to users). 

Below a quick sketch that I've made of the data table I would like to be created and updated in the backend. Regarding the last column: I've realised that we will in fact need to gather the texts themselves—at least for the research context—in order for us to be able to monitor and moderate. 

![IMG_7088](https://github.com/Gryphire/wisperbot/assets/17269537/eb7c7d4f-b7e8-4566-8e61-d56c79d979f9)

**Comment by Gryphire** [comment]
Awesome. As long as I can easily get the data in table format (as sketched), I'm happy.

**Comment by lydgate** [comment]
Just looking at this backend data structuring. starting to come together. One thing is, for me it always logs that I've chosen a prompt, when really it is retrieving a stored prompt. Just wondered how you want that reflected in the log? As in... do you want 2 different event types depending on if a prompt is newly chosen or retrieved? Or the same event type? Or does it only need to be logged when a new one is chosen?

Here's a first attempt:

```csv
2023-09-21 17:23:07.544139,5920741802,Bryan,sent text to Wisper,hi
2023-09-21 17:23:07.544268,5920741802,Bryan,chose/retrieved prompt,Authentic Relating
2023-09-21 17:23:07.693738,5920741802,Wisperbot,sent response,Hi there! Welcome to WisperBot. Please use the /start command to get instructions on how to get started!
```

**Comment by lydgate** [comment]
Another question @Gryphire: Do you want the full bot text to appear in the logging too?

Option 1:
`2023-09-21 20:34:21.133171,5920741802,Wisperbot,Sent help info to Bryan,`

Option 2:
```
2023-09-21 20:34:53.903216,5920741802,Wisperbot,Sent help info to Bryan,"I'm happy to tell you some more about WisperBot`

WisperBot is a product of the Games for Emotional and Mental Health Lab, and has been created to facilitate asynchronous audio conversations between you and others around topics that matter to you.

Through the prompts that WisperBot provides, the bot's purpose is to help you connect with others in a meaningful way.

Not sure how to get started? Use the /start command to review the instructions."
```

**Comment by Gryphire** [comment]
Hmm, good question.

I guess the answer depends on this: is the logging (as it is done to console) the same as the data logging that’s done to the backend database? Because I don’t need all of that info when it comes to logging to the console, but when it comes to logging to the backend data, then yes that’s going to be useful for the research context (i.e., in case I need to figure out post-hoc what might have happened).

If those two are one and the same, then yes let’s keep it because the backend data being comprehensive is most important. (:

**Comment by lydgate** [comment]
Just to clarify, I've modified the log method of the chat class (i.e., chat.log) to both log to the console and also to save to a csv datastore.

A few questions:

1. Is it ok to log "sent text" instead of "sent text to Wisper"? Just because they could also send it to a group.
1. Do you also want "backend" actions to appear in the backend data? For example, when the bot does transcription. Or do you only want stuff that happens in the chat to be logged?
1. Do you need emojis to be logged? That's not currently done and might be a bit trickier.

**Comment by Gryphire** [comment]
Tried out the #16 branch, here are my answers + a couple of minor things (I can go in an change them myself but I want to avoid interfering with whatever changes you're already making on your end).

Answers:
1. So _ideally_I would want this to be separated out (which I'm hoping might be possible since the bot has built-in logic for handling messages that contain its name—and are therefore considered to be addressing the bot—versus messages that don't). In the scenario that there's a lot of text messages sent around it'd be good to be able to see in a split second whom someone's trying to talk to (rather than having to search through the text to establish whether it was intended for Wisper or for the other chat members). 
2. Yes, the way it's being done now is totally fine (i.e., the bot's downloading and transcribing of voice notes actions being included in the data file)
3. As far as I can see, emojis _are_ being logged in the text, right? If I import the CSV into Excel with UTF-8 encoding defined they show up alright. Or do you mean emoji reactions to messages? Those don't need to be logged right now, I don't foresee any issues with that. 

Couple of observations:
1. We'll need to switch the separator for the CSV to something else (I think semicolon should work out) since parsing issues occur when handling message texts (they get chopped up by the comma). 
2. Minor point: 'Sent new prompt' / 'Sent help info' and similar events where user uses one of the / commands, let's rephrase this to the user's perspective since that makes most sense in the stream of events. 'Sent new prompt' would thus instead be '[username] requested new prompt with /prompt', 'sent help info' would be '[username] requested help info with /help', etc. That way, when reading back the data it makes chronological sense. (see screenshot for example of prompt-requesting sequence)
<img width="1207" alt="image" src="https://github.com/Gryphire/wisperbot/assets/17269537/dfe1f3e2-e833-45a5-9f48-57bb340f0b69">
3. Last, all the data from different chats is now being logged to a general data file as opposed to a chat-specific data file. Just want to make sure you're aware!

Again, more than happy to do fixes 1 and 2 myself but I'm wary of our beams crossing and screwing up the versioning, so let me know!

**Comment by lydgate** [comment]
Thanks Nastasia!

1. OK, I think because I haven't seen any of the actual text people send, I don't really understand how to categorize who is being addressed. It could be categorized if they use @username, but I was coding on the assumption that people mostly wouldn't do that, so I'm not sure how it will determine "to whom" they are speaking. Basically I don't understand how to programmatically determine who is being spoken to if there is no @ in the message. Feel free to work on this in the branch if you have an idea though!
1. OK great!
1. Yes, you're correct, emojis within the text should be logged. I was asking about emoji reactions to texts. Those are not logged (and it sounds like they don't need to be!)

Observations:

1. The csv currently should use double quotes when a comma is contained, and I think on import to Excel or Google Sheets you should be able to take this into account? Or I could put everything into quotes if that helps; let me know. Obviously semicolons can have the same issues if people use them in text.
1. OK, will take a look at this soon :)
1. Ah, did you want each chat to log to a separate csv in the private folders? That can also be done. I thought you were saying you wanted it all logged centrally. Let me know :)

I won't have time to work on this until Tuesday at the earliest, so feel free to take a stab at 1-2 if you have time before then (but obviously no worries if not!) and I will pull down any changes you make to this branch before proceeding.

**Comment by Gryphire** [comment]
Ok, I've made some changes but am not able to test them before tomorrow to see if they work out the way I expect, in particular with regards to the message parsing (point 2). It seems to have to do not with the comma separation but with the line breaks that are contained within certain texts (like the prompt texts, and the start and help info). I don't want to remove those from the messages since they make the content a lot easier to read for users, but I've now tried to strip them from those line breaks when they're being saved to log (using the strip() method). 

I've also made some minor adjustments to the phrasing of the logs as per point 1. 

With regards to point 3 of the observations; yes, we need the data to be saved separately per chat, just like we had when we just had the log. ^^

**Comment by lydgate** [comment]
OK, understood! I can take a look at separating them out later this week.

I think since there is a column for the chat ID, and the folder hierarchy already contains the chat ID, I thought that you wanted them all in one log (so that you could see exactly what the bot was doing across multiple conversations). I now understand that you want many CSVs and not one.

**Comment by lydgate** [comment]
Hi Nastasia,

OK, the latest change on this branch should log to individual CSV files. Please note that this is **instead** of the top-level CSV file. Let me know if you actually wanted it logged into **both** individual files and a top-level file.

I have not yet made changes to detect "whom" someone is speaking to.

It looks like you rephrased some of the logging from "Sent new prompt to <person>" to "<person> requested new prompt," but let me know if any further rephrasing is required.

Maybe worth you testing the latest version then we can chat to refresh about any remaining issues?

Thanks!

Bryan

**Comment by Gryphire** [comment]
Ran into an error to do with the 'original_prompt' variable when trying to interact with the bot, see error below (and the bot consequently doesn't respond appropriately). The / commands are not affected, only seems to happen when sending text to the bot. (:

`2023-11-21 12:04:08,513 - chat-5625099873 - INFO - Nastasia Sent text: Hey
2023-11-21 12:04:08,514 - telegram.ext.Application - ERROR - No error handlers are registered, logging exception.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/telegram/ext/_application.py", line 1124, in process_update
    await coroutine
  File "/usr/local/lib/python3.9/site-packages/telegram/ext/_handler.py", line 141, in handle_update
    return await self.callback(update, context)
  File "/Users/nastasia/surfdriveUT/wisperbot/wisperbot_21112023_16branch.py", line 305, in handle_msg
    text=create_response(chat,usertext)
  File "/Users/nastasia/surfdriveUT/wisperbot/wisperbot_21112023_16branch.py", line 272, in create_response
    if chat.prompt != original_prompt:
UnboundLocalError: local variable 'original_prompt' referenced before assignment`

**Comment by lydgate** [comment]
Oops! I think let's set original_prompt to a null string at the start and see if that fixes it? I've pushed a change that does that. 20e30c70be4930ad90e987f98b389eae56291719

Can you let me know if that works?

Bryan

**Comment by Gryphire** [comment]
Hi hi!

We've got this popping up now hahaha!

`2024-01-05 17:26:03,045 - chat-5625099873 - INFO - Nastasia Sent text: Hey
2024-01-05 17:26:03,081 - telegram.ext.Application - ERROR - No error handlers are registered, logging exception.
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/telegram/ext/_application.py", line 1173, in process_update
    await coroutine
  File "/usr/local/lib/python3.11/site-packages/telegram/ext/_basehandler.py", line 141, in handle_update
    return await self.callback(update, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/u278158/surfdrive2/wisperbot/wisperbot_05012024_16branch.py", line 306, in handle_msg
    text=create_response(chat,usertext)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/u278158/surfdrive2/wisperbot/wisperbot_05012024_16branch.py", line 274, in create_response
    chat.log(f"{chat.name} prompt is {chat.prompt}", group_member=chat.name, event='chose/retrieved prompt', content=chat.prompt.strip())
                                                                                                                     ^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'strip'`
